Backend/Remora.Discord.Voice/DiscordVoiceClient.cs

```diff
@@ -0,0 +1,1258 @@
+        /// <param name="connectionWaiterService">The connection waiter service.</param>
+        /// <param name="transportService">The voice payload transport service.</param>
+        /// <param name="dataService">The voice data transport service.</param>
+        /// <param name="defaultTranscoder">The default audio transcoder to use.</param>
+        /// <param name="random">A random number generator.</param>
+        public DiscordVoiceClient
+        (
+            ILogger<DiscordVoiceClient> logger,
+            DiscordGatewayClient gatewayClient,
+            IOptions<DiscordVoiceClientOptions> clientOptions,
+            IConnectionEstablishmentWaiterService connectionWaiterService,
+            IVoicePayloadTransportService transportService,
+            IVoiceDataTranportService dataService,
+            IAudioTranscoderService defaultTranscoder,
+            Random random
+        )
+        {
+            _logger = logger;
+            _gatewayClient = gatewayClient;
+            _clientOptions = clientOptions.Value;
+            _connectionWaiterService = connectionWaiterService;
+            _transportService = transportService;
+            _dataService = dataService;
+            _random = random;
+
+            _heartbeatData = new HeartbeatData();
+            _payloadsToSend = new ConcurrentQueue<IVoicePayload>();
+            _receivedPayloads = new ConcurrentQueue<IVoicePayload>();
+
+            _disconnectRequestedSource = new CancellationTokenSource();
+            _runLoopTask = Task.FromResult(Result.FromSuccess());
+            _sendTask = Task.FromResult(Result.FromSuccess());
+            _receiveTask = Task.FromResult(Result.FromSuccess());
+
+            _transmitSemaphore = new SemaphoreSlim(1, 1);
+            _defaultTranscoder = defaultTranscoder;
+
+            _silenceFrameBuffer = MemoryPool<byte>.Shared.Rent(3);
+            _silenceFrameBuffer.Memory.Span[0] = 0xF8;
+            _silenceFrameBuffer.Memory.Span[0] = 0xFF;
+            _silenceFrameBuffer.Memory.Span[0] = 0xFE;
+
+            this.ConnectionStatus = GatewayConnectionStatus.Offline;
+        }
+
+        /// <summary>
+        /// Runs the voice gateway client.
+        /// <para>
+        /// This task will not complete until cancelled (or faulted), maintaining the connection for the duration of it.
+        ///
+        /// If the gateway client encounters a fatal problem during the execution of this task, it will return with a
+        /// failed result. If a shutdown is requested, it will gracefully terminate the connection and return a
+        /// successful result.
+        /// </para>
+        /// </summary>
+        /// <param name="guildID">The ID of the guild to connect to.</param>
+        /// <param name="channelID">The ID of the channel to connect to.</param>
+        /// <param name="isSelfMuted">A value indicating whether the bot should mute itself.</param>
```

Comment:
Perhaps I'm nitpicking here, but shouldn't this be along the lines of `Whether the user joins in a muted state`? 

`[...] in a self-muted state` would also work, but perhaps would be a bit long for the documentation.

Also, given this context, I think `joinSelfMuted` would be a more apt parameter name, given what the documentation says it does.

Reply:
I prefer that description, thanks. I've gone with the below for now.
> A value indicating whether the bot should connect in a muted state.

Given that it's a persistent state till changed, and not just for the period of joining the call, the current name makes more sense to me though.